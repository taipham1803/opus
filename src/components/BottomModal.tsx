import React from 'react';
import {StyleSheet, View, ViewStyle, Dimensions} from 'react-native';
import {
  PanGestureHandler,
  PanGestureHandlerGestureEvent,
} from 'react-native-gesture-handler';
import Animated, {
  Easing,
  interpolate,
  useAnimatedGestureHandler,
  useAnimatedStyle,
  useDerivedValue,
  useSharedValue,
  withSpring,
  withTiming,
  WithSpringConfig,
  WithTimingConfig,
} from 'react-native-reanimated';

const screen = Dimensions.get('screen');

export type BottomModalProps = {
  height: number;
  isOpen?: boolean;
  backdropColor?: string;
  style?: ViewStyle;
  animation?: 'spring' | 'timing';
  springConfig?: WithSpringConfig;
  timingConfig?: WithTimingConfig;
  backdropStyle?: ViewStyle;
  children: JSX.Element[];
};

export const BottomModal = ({
  height,
  isOpen,
  backdropColor,
  style,
  children,
  backdropStyle,
  animation,
  springConfig,
  timingConfig,
}: BottomModalProps) => {
  const topAnim = useSharedValue(screen.height);
  const isMounted = React.useRef<boolean>(false);

  const updateTop = React.useCallback(
    (value: number) => {
      'worklet';
      if (animation === 'spring') {
        return withSpring(value, springConfig);
      } else {
        return withTiming(value, timingConfig);
      }
    },
    [timingConfig, animation, springConfig],
  );

  React.useEffect(() => {
    if (isMounted.current) {
      console.log('isOpen: ' + isOpen);
      if (isOpen) {
        // topAnim.value = updateTop(screen.height);
        // topAnim.value = updateTop(screen.height - height);
      } else {
        // topAnim.value = updateTop(screen.height);
      }
    } else {
      isMounted.current = true;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isOpen]);

  const isActive = useDerivedValue<boolean>(() => {
    if (topAnim.value > screen.height - 10) {
      return false;
    } else {
      return true;
    }
  }, [topAnim]);

  const gestureHandler = useAnimatedGestureHandler<
    PanGestureHandlerGestureEvent,
    {startHeight: number}
  >({
    onStart: (_, context) => {
      context.startHeight = topAnim.value;
    },
    onActive: (event, context) => {
      if (context.startHeight + event.translationY > screen.height - height) {
        topAnim.value = context.startHeight + event.translationY;
      }
    },
    onEnd: () => {
      if (topAnim.value > screen.height - height / 2) {
        topAnim.value = updateTop(screen.height);
      } else {
        topAnim.value = updateTop(screen.height - height);
      }
    },
  });

  const containerAnimatedStyle = useAnimatedStyle(() => ({
    top: topAnim.value,
  }));

  const backdropAnimatedStyle = useAnimatedStyle(() => ({
    opacity: interpolate(
      topAnim.value,
      [screen.height - height, screen.height],
      [1, 0],
    ),
    top: isActive.value ? 0 : screen.height,
  }));

  return (
    <View style={styles.fullScreen}>
      <Animated.View
        style={[
          styles.backdrop,
          {backgroundColor: backdropColor},
          backdropStyle,
          backdropAnimatedStyle,
        ]}
      />
      <PanGestureHandler onGestureEvent={gestureHandler}>
        <Animated.View
          style={[styles.container, {height}, style, containerAnimatedStyle]}>
          {children}
        </Animated.View>
      </PanGestureHandler>
    </View>
  );
};

BottomModal.defaultProps = {
  timingConfig: {duration: 300, easing: Easing.quad},
  animation: 'timing',
};

const styles = StyleSheet.create({
  fullScreen: {
    height: screen.height,
    position: 'absolute',
    zIndex: 10,
    top: 0,
    left: screen.width / 2,
    justifyContent: 'center',
    alignSelf: 'center',
    alignItems: 'center',
  },
  backdrop: {
    position: 'absolute',
    height: screen.height,
    width: screen.width,
  },
  container: {
    width: screen.width,
    position: 'absolute',
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'white',
    borderTopLeftRadius: 20,
    borderTopRightRadius: 20,

    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 4,
    },
    shadowOpacity: 0.3,
    shadowRadius: 4.65,
    elevation: 8,
  },
});
