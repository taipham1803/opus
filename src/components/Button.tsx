import React from 'react';
import {Text, TouchableOpacity, ViewStyle} from 'react-native';
import {Colors, Fonts, Style} from 'styles';

export type ButtonAppearance =
  | 'primary'
  | 'secondary'
  | 'borderFull'
  | 'borderLess'
  | 'danger';

interface ButtonProps {
  title?: string;
  style?: ViewStyle[];
  appearance?: ButtonAppearance;
  disabled?: boolean;
  onPress?: () => void;
  borderWidth?: number;
}

export const getButtonAppearanceStyles = ({
  appearance = 'primary',
}: ButtonProps) => {
  const styles = [];

  if (appearance === 'primary') {
    styles.push(Style.con({bg: Colors.interactive.stroke}));
  } else if (appearance === 'secondary') {
    styles.push(Style.con({bg: Colors.tran}));
  } else if (appearance === 'borderFull') {
    styles.push(Style.con({bg: Colors.interactive.stroke, py: 6, bor: 99}));
  } else if (appearance === 'borderLess') {
    styles.push(Style.con({bg: Colors.interactive.stroke}));
    styles.push(Style.border({bor: 5, width: 0}));
  }

  return styles;
};

export const Button = ({
  title,
  style,
  appearance = 'primary',
  disabled = false,
  onPress = () => {},
  borderWidth = 2,
}: ButtonProps) => {
  const passedInProps = {
    appearance,
    disabled,
  };
  const appearanceStyles = getButtonAppearanceStyles(passedInProps);
  return (
    <TouchableOpacity
      onPress={onPress}
      style={[
        Style.con({py: 14, cen: true}),
        Style.border({bor: 5, width: borderWidth, color: Colors.primary}),
        ...appearanceStyles,
        style,
      ]}>
      <Text style={Fonts.t(15, Colors.blue[2], {wei: '500'})}>{title}</Text>
    </TouchableOpacity>
  );
};
