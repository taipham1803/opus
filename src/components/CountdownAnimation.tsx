import React from 'react';
import {Text, TextStyle, View, ViewStyle} from 'react-native';
import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  withRepeat,
  runOnJS,
} from 'react-native-reanimated';
import {Colors, Fonts, Style} from 'styles';

interface CountdownAnimationProps {
  style?: ViewStyle;
  numberStyle?: TextStyle;
  onAnimationEnded?: () => void;
  description?: string;
  descriptionStyle?: TextStyle;
  countdownFrom?: number;
  stopCountOn?: number;
}

export const CountdownAnimation = ({
  style,
  numberStyle,
  onAnimationEnded,
  description,
  descriptionStyle,
  countdownFrom = 3,
  stopCountOn = 0,
}: CountdownAnimationProps) => {
  const progress = useSharedValue(1);
  const scale = useSharedValue(1);
  const [countDown, setCountDown] = React.useState<number>(countdownFrom);

  const reanimatedStyle = useAnimatedStyle(() => {
    return {
      opacity: progress.value,
      transform: [{scale: scale.value}],
    };
  }, []);

  const updateNumber = React.useCallback(() => {
    if (countDown > stopCountOn) {
      setCountDown(pre => pre - 1);
    }
  }, [countDown, stopCountOn]);

  const callbackEndAnimation = () => onAnimationEnded && onAnimationEnded();

  React.useEffect(() => {
    progress.value = withRepeat(
      withTiming(0, {duration: 1500}, (isFinished?: boolean) => {
        if (isFinished) {
          runOnJS(updateNumber)();
        }
      }),
      3,
      false,
      (isFinished?: boolean) => {
        if (isFinished) {
          runOnJS(callbackEndAnimation)();
        }
      },
    );
    scale.value = withRepeat(withTiming(15, {duration: 1500}), 3, false);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <View style={Style.con({flex: 1, bg: Colors.primary})}>
      <View style={[Style.con({flex: 1, cen: true}), style]}>
        <Animated.Text
          style={[
            Fonts.t(21, Colors.black, {wei: '700'}),
            numberStyle,
            reanimatedStyle,
          ]}>
          {countDown}
        </Animated.Text>
      </View>
      {!!description && (
        <Text
          style={[
            Fonts.t(21, Colors.gray[1], {
              wei: '700',
              self: 'center',
              text: 'center',
              b: 44,
            }),
            descriptionStyle,
          ]}>
          {description}
        </Text>
      )}
    </View>
  );
};
