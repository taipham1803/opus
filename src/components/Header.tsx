import {useNavigation} from '@react-navigation/core';
import {icons} from 'assets';
import React from 'react';
import {View, Text, TouchableOpacity, ViewStyle, TextStyle} from 'react-native';
import FastImage from 'react-native-fast-image';
import {DrawerActions} from '@react-navigation/native';

import {Colors, Fonts, Style} from 'styles';

export type leftAppearance = 'back' | 'drawer';

interface HeaderProps {
  title?: string;
  titleUpperCase?: boolean;
  titleStyle?: TextStyle;
  style?: ViewStyle;
  leftAppearance?: leftAppearance;
  onLeft?: () => void;
}

export const getIconLeftAppearance = (appearance: leftAppearance) => {
  if (appearance === 'back') {
    return icons.iconChevronLeft;
  } else if (appearance === 'drawer') {
    return icons.iconMenu;
  }
};

export const Header = ({
  title,
  titleUpperCase = true,
  titleStyle,
  style,
  leftAppearance,
}: HeaderProps) => {
  const nav = useNavigation();
  const onPressLeft = () => {
    if (leftAppearance === 'back') {
      nav.goBack();
    } else if (leftAppearance === 'drawer') {
      nav.dispatch(DrawerActions.toggleDrawer());
    }
  };
  return (
    <View
      style={[
        Style.con({direc: 'row', h: 44, items: 'center'}),
        Style.border({bbw: 1, color: Colors.gray.light}),
        style,
      ]}>
      <TouchableOpacity
        onPress={onPressLeft}
        style={Style.con({w: 60, pl: 12, justify: 'center'})}>
        {!!leftAppearance && (
          <FastImage
            style={Style.con({size: 24})}
            source={getIconLeftAppearance(leftAppearance)}
          />
        )}
      </TouchableOpacity>
      <View style={Style.con({flex: 1, cen: true})}>
        <Text
          style={[
            Fonts.t(15, Colors.blue[2], {wei: '700', text: 'center'}),
            titleStyle,
          ]}>
          {titleUpperCase ? title?.toUpperCase() : title}
        </Text>
      </View>
      <View style={Style.con({w: 60})} />
    </View>
  );
};
