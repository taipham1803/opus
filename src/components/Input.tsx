import {icons} from 'assets';
import React from 'react';
import {View, Text, TextInput, ViewStyle, TouchableOpacity} from 'react-native';
import FastImage from 'react-native-fast-image';
import {Colors, Fonts, Style} from 'styles';

interface InputProps {
  keyName?: string;
  title?: string;
  style?: ViewStyle;
  secure?: boolean;
  disabled?: boolean;
  placeholder?: string;
  onChange?: (keyName: string, value: string) => void;
}

export const Input = ({
  keyName,
  title,
  style,
  secure = false,
  disabled = false,
  placeholder = '',
  onChange,
}: InputProps) => {
  const [isShowPass, setShowPass] = React.useState<boolean>(false);
  const onPressEyeIcon = () => setShowPass(preState => !preState);
  const onChangeText = React.useCallback(
    _value => {
      onChange && keyName && onChange(keyName, _value);
    },
    [onChange, keyName],
  );
  return (
    <View style={[style]}>
      <Text style={[Fonts.t(12, Colors.blue[1], {wei: '600', b: 8})]}>
        {title}
      </Text>
      <View style={Style.con({direc: 'row'})}>
        <TextInput
          editable={!disabled}
          placeholder={placeholder}
          style={[Fonts.t(14, Colors.inactive.text), Style.con({flex: 1})]}
          clearButtonMode="while-editing"
          secureTextEntry={secure && !isShowPass}
          onChangeText={onChangeText}
        />
        {!!secure && (
          <TouchableOpacity
            style={Style.con({size: 20, cen: true, ml: 8})}
            onPress={onPressEyeIcon}>
            <FastImage
              style={Style.con({size: 16})}
              source={isShowPass ? icons.iconVisible : icons.iconInvisible}
            />
          </TouchableOpacity>
        )}
      </View>
      <View style={Style.con({h: 1, bg: Colors.gray[2], mt: 8})} />
    </View>
  );
};
