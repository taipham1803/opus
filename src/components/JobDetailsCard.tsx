import React from 'react';
import {Text, View, ViewStyle} from 'react-native';

import {JobProperty} from 'entity/suggest';
import {Colors, Fonts, Style} from 'styles';

interface JobDetailsCardProps {
  style?: ViewStyle;
  properties: JobProperty[];
  title?: string;
}

export const JobDetailsCard = ({
  properties,
  style,
  title,
}: JobDetailsCardProps) => {
  return (
    <View style={[Style.con({mt: 8}), style]}>
      <View style={Style.con({mx: 16, bg: Colors.gray.light, h: 1})} />
      <View style={Style.con({p: 16})}>
        <Text style={Fonts.t(16, Colors.blue[1], {wei: '500', t: 8, b: 16})}>
          {title ?? 'Details'.toUpperCase()}
        </Text>
        {Array.isArray(properties) &&
          properties.map((property: JobProperty) => {
            if (!property.icon || !property.title) {
              return null;
            }
            return (
              <View
                style={Style.con({direc: 'row', pt: 4, pb: 8})}
                key={property.id}>
                <View style={Style.con({w: 80, mr: 20})}>
                  <Text
                    style={Fonts.t(
                      property.fontSize ?? 12,
                      property.color ?? Colors.blue[1],
                      {
                        wei: property.fontWeight ?? '500',
                      },
                    )}>
                    {property.title}
                  </Text>
                </View>
                <Text
                  style={Fonts.t(
                    property.fontSize ?? 14,
                    property.color ?? Colors.blue[1],
                    {
                      wei: property.fontWeight ?? '600',
                    },
                  )}>
                  {property.content}
                </Text>
              </View>
            );
          })}
      </View>
    </View>
  );
};
