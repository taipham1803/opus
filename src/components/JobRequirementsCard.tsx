import React from 'react';
import {Text, View, ViewStyle} from 'react-native';

import {JobRequirement} from 'entity/suggest';
import {Colors, Fonts, Style} from 'styles';

interface JobRequirementsCardProps {
  style?: ViewStyle;
  requirements: JobRequirement[];
}

export const JobRequirementsCard = ({
  requirements,
  style,
}: JobRequirementsCardProps) => {
  return (
    <View style={[Style.con({mt: 8}), style]}>
      <View style={Style.con({mx: 16, bg: Colors.gray.light, h: 1})} />
      <View style={Style.con({p: 16})}>
        <Text style={Fonts.t(16, Colors.blue[1], {wei: '500', t: 8, b: 16})}>
          {'Requirements'.toUpperCase()}
        </Text>
        {Array.isArray(requirements) &&
          requirements.map((property: JobRequirement) => {
            if (!property.title) {
              return null;
            }
            return (
              <View
                style={Style.con({direc: 'row', pt: 4, pb: 8})}
                key={property.id}>
                <View style={Style.con({w: 80, mr: 20})}>
                  <Text
                    style={Fonts.t(
                      property.fontSize ?? 12,
                      property.color ?? Colors.blue[1],
                      {
                        wei: property.fontWeight ?? '500',
                      },
                    )}>
                    {property.title}
                  </Text>
                </View>
                <Text
                  style={Fonts.t(
                    property.fontSize ?? 14,
                    property.color ?? Colors.blue[1],
                    {
                      wei: property.fontWeight ?? '600',
                    },
                  )}>
                  {property.value.join(', ')}
                </Text>
              </View>
            );
          })}
      </View>
    </View>
  );
};
