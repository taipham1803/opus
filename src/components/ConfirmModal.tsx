import React from 'react';
import {Text, ViewStyle} from 'react-native';
import Modalbox from 'react-native-modalbox';
import {Colors, Fonts, Style} from 'styles';
import {Button} from './Button';

interface ConfirmModalProps {
  style?: ViewStyle;
  open: boolean;
  onClosed: () => void;
  modalHeight: number;
}

export const ConfirmModal = ({
  style,
  open,
  onClosed,
  modalHeight = 200,
}: ConfirmModalProps) => {
  return (
    <Modalbox
      style={[
        Style.con({
          w: '100%',
          h: modalHeight,
          over: 'hidden',
          bg: Colors.white,
          p: 20,
        }),
        Style.border({btlr: 16, btrr: 16}),
        style,
      ]}
      useNativeDriver={true}
      backdropPressToClose={true}
      backdropColor={'rgba(16, 15, 15, 0.7)'}
      animationDuration={500}
      backdrop={true}
      backButtonClose={true}
      isOpen={open}
      swipeToClose={true}
      position={'bottom'}
      coverScreen={true}
      onClosed={onClosed}>
      <Text style={Fonts.t(16, Colors.blue[1], {wei: '600'})}>
        {'Your request has been submitted'}
      </Text>
      <Text style={Fonts.t(14, Colors.blue[1], {t: 12})}>
        {'Your recruiter will contact you shortly.'}
      </Text>
      <Button
        title={'ok'.toUpperCase()}
        style={Style.con({mt: 24})}
        onPress={onClosed}
        appearance="borderLess"
      />
    </Modalbox>
  );
};
