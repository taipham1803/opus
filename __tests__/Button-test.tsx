import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {Button} from 'components/Button';

describe('Button', () => {
  describe('Rendering', () => {
    it('should match to snapshot', () => {
      const component = renderer.create(<Button title={'Sign in!'} />).toJSON();
      expect(component).toMatchSnapshot();
    });
  });
});

// describe('onPressHandler', () => {
//   it('should call onPress', () => {
//     const mockOnPress = jest.fn(); // 1. mock function
//     const component = renderer.create(
//       <Button
//         title="test label"
//         onPress={mockOnPress} // 2. passing in mock function as props
//       />,
//     );
//     // 3. getting an instance of component
//     const instance: any = component.getInstance();
//     // 4. manually triggering onPressHandler()
//     instance?.onPressHandler();
//     // Act
//     expect(mockOnPress).toHaveBeenCalled();
//   });
// });
