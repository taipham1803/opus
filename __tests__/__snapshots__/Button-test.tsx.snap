// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Rendering should match to snapshot 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#E3F8F4",
      "borderColor": "#99E1E1",
      "borderRadius": 10,
      "borderWidth": 4,
      "bottom": undefined,
      "justifyContent": "center",
      "left": undefined,
      "opacity": 1,
      "paddingVertical": 23,
      "right": undefined,
      "top": undefined,
    }
  }
>
  <Text
    style={
      Object {
        "alignSelf": "auto",
        "color": "#03606A",
        "fontSize": 15,
        "fontWeight": "500",
        "marginBottom": undefined,
        "marginHorizontal": undefined,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "marginVertical": undefined,
        "opacity": undefined,
        "textAlign": "auto",
        "textAlignVertical": "auto",
      }
    }
  >
    Sign in!
  </Text>
</View>
`;
